File: src/ui/pickers.ts
/**
 * 编辑表单里的各 picker（供应商 / API 地址 / 认证字段 / effort），对齐现版 Pick-*。
 */
import { getProviderEnvMap, type Store } from '../config/store.js';
import type { Preset } from '../config/types.js';
import { T } from '../i18n/index.js';
import { padDisplay } from '../utils/display.js';
import { selectMenu } from './select.js';
import { readText } from './text.js';

/** 选供应商：从目录选一个 / 自定义手填名 / 不改。 */
export async function pickProvider(catalog: Preset[], current: string): Promise<Preset | 'custom' | null> {
  const names = catalog.map((p) => p.name);
  const items = [...names, T('pick.provider.custom'), T('pick.noChange')];
  const sel = await selectMenu({
    title: T('pick.provider.title', current || T('pick.provider.none')),
    items,
    hint: T('pick.hint'),
  });
  if (sel < 0 || sel === items.length - 1) return null;
  if (sel === names.length) return 'custom';
  return catalog[sel] ?? null;
}

/** 供应商有多个地址时让用户选一个；只有一个直接用，无地址保持原值。 */
export async function pickProviderUrl(preset: Preset, current: string): Promise<string> {
  const urls = preset.urls;
  if (urls.length === 0) return current;
  if (urls.length === 1) return urls[0]?.url ?? current;
  const labels = urls.map((u) => `${padDisplay(u.label, 12)} ${u.url || T('empty.paren')}`);
  const items = [...labels, T('pick.noChange')];
  const sel = await selectMenu({ title: T('pick.providerUrl.title', preset.name), items, hint: T('pick.hint') });
  if (sel < 0 || sel === items.length - 1) return current;
  return urls[sel]?.url ?? current;
}

/** 选 API 地址：目录所有 url + 已有配置用过的 url + 手动输入 + 不修改。 */
export async function pickBaseUrl(current: string, store: Store, catalog: Preset[]): Promise<string> {
  const entries: Array<{ label: string; url: string }> = [];
  const seen = new Set<string>();
  for (const p of catalog) {
    for (const u of p.urls) {
      const tag = p.urls.length > 1 ? `${p.name}/${u.label}` : p.name;
      entries.push({ label: `${padDisplay(tag, 20)} ${u.url || T('empty.paren')}`, url: u.url });
      seen.add(u.url);
    }
  }
  for (const prov of store.providers) {
    const u = getProviderEnvMap(prov).ANTHROPIC_BASE_URL;
    if (u && u.trim() && !seen.has(u)) {
      seen.add(u);
      entries.push({ label: `${padDisplay(T('pick.base.existing', prov.name), 20)} ${u}`, url: u });
    }
  }
  const items = [...entries.map((e) => e.label), T('pick.manual'), T('pick.noChange')];
  const sel = await selectMenu({ title: T('pick.base.title', current || T('empty.paren')), items, hint: T('pick.hint') });
  if (sel < 0 || sel === items.length - 1) return current;
  if (sel < entries.length) return entries[sel]?.url ?? current;
  const v = await readText(`  ${T('pick.base.manualInput')}`);
  if (v === undefined || v === '') return current;
  if (v === '-') return '';
  return v.trim();
}

/** 选认证字段：AUTH_TOKEN / API_KEY / 不改。 */
export async function pickAuth(current: 'AUTH_TOKEN' | 'API_KEY'): Promise<'AUTH_TOKEN' | 'API_KEY'> {
  const items = [T('pick.auth.token'), T('pick.auth.apikey'), T('pick.noChange')];
  const sel = await selectMenu({ title: T('pick.auth.title', current), items, hint: T('pick.hint') });
  if (sel === 0) return 'AUTH_TOKEN';
  if (sel === 1) return 'API_KEY';
  return current;
}

const EFFORT_OPTS = ['low', 'medium', 'high', 'xhigh', 'max', 'auto'];

/** 选 effort 思考档：low…auto / 留空 / 不改。 */
export async function pickEffort(current: string): Promise<string> {
  const items = [...EFFORT_OPTS, T('pick.effort.empty'), T('pick.noChange')];
  const sel = await selectMenu({
    title: T('pick.effort.title', current || T('empty.paren')),
    items,
    hint: T('pick.effort.hint'),
  });
  if (sel < 0 || sel === items.length - 1) return current;
  if (sel === EFFORT_OPTS.length) return '';
  return EFFORT_OPTS[sel] ?? current;
}

File: internal/tui/pickers.go
package tui

import (
	"strings"

	"github.com/becomeless/cc-x/internal/config"
	"github.com/becomeless/cc-x/internal/display"
	"github.com/becomeless/cc-x/internal/i18n"
	"github.com/becomeless/cc-x/internal/presets"
)

func orEmpty(s string) string {
	if s == "" {
		return i18n.T("empty.paren")
	}
	return s
}

// PickProvider 选供应商：返回 (选中供应商, custom)。两者皆零值表示不改。
func PickProvider(t *Terminal, catalog []presets.Preset, current string) (sel *presets.Preset, custom bool) {
	names := make([]string, len(catalog))
	for i := range catalog {
		names[i] = catalog[i].Name
	}
	items := append(append([]string{}, names...), i18n.T("pick.provider.custom"), i18n.T("pick.noChange"))
	cur := current
	if cur == "" {
		cur = i18n.T("pick.provider.none")
	}
	s := SelectMenu(t, SelectOptions{Title: i18n.T("pick.provider.title", cur), Items: items, Hint: i18n.T("pick.hint"), NoNumber: true})
	if s < 0 || s == len(items)-1 {
		return nil, false
	}
	if s == len(names) {
		return nil, true
	}
	return &catalog[s], false
}

// PickProviderURL 供应商有多个地址时让用户选一个；只有一个直接用，无地址保持原值。
func PickProviderURL(t *Terminal, preset *presets.Preset, current string) string {
	urls := preset.URLs
	if len(urls) == 0 {
		return current
	}
	if len(urls) == 1 {
		return urls[0].URL
	}
	labels := make([]string, len(urls))
	for i, u := range urls {
		labels[i] = display.Pad(u.Label, 12) + " " + orEmpty(u.URL)
	}
	items := append(labels, i18n.T("pick.noChange"))
	s := SelectMenu(t, SelectOptions{Title: i18n.T("pick.providerUrl.title", preset.Name), Items: items, Hint: i18n.T("pick.hint"), NoNumber: true})
	if s < 0 || s == len(items)-1 {
		return current
	}
	return urls[s].URL
}

// PickBaseURL 选 API 地址：目录所有 url + 已有配置用过的 url + 手动输入 + 不修改。
func PickBaseURL(t *Terminal, current string, store *config.Store, catalog []presets.Preset) string {
	type entry struct{ label, url string }
	var entries []entry
	seen := map[string]bool{}
	for i := range catalog {
		p := catalog[i]
		for _, u := range p.URLs {
			tag := p.Name
			if len(p.URLs) > 1 {
				tag = p.Name + "/" + u.Label
			}
			entries = append(entries, entry{label: display.Pad(tag, 20) + " " + orEmpty(u.URL), url: u.URL})
			seen[u.URL] = true
		}
	}
	for i := range store.Providers {
		u := config.GetProviderEnvMap(store.Providers[i])["ANTHROPIC_BASE_URL"]
		if u != "" && !seen[u] {
			seen[u] = true
			entries = append(entries, entry{label: display.Pad(i18n.T("pick.base.existing", store.Providers[i].Name), 20) + " " + u, url: u})
		}
	}
	items := make([]string, 0, len(entries)+2)
	for _, e := range entries {
		items = append(items, e.label)
	}
	items = append(items, i18n.T("pick.manual"), i18n.T("pick.noChange"))
	s := SelectMenu(t, SelectOptions{Title: i18n.T("pick.base.title", orEmpty(current)), Items: items, Hint: i18n.T("pick.hint"), NoNumber: true})
	if s < 0 || s == len(items)-1 {
		return current
	}
	if s < len(entries) {
		return entries[s].url
	}
	v, ok := ReadText(t, "  "+i18n.T("pick.base.manualInput"))
	if !ok || v == "" {
		return current
	}
	if v == "-" {
		return ""
	}
	return strings.TrimSpace(v)
}

// PickAuth 选认证字段：AUTH_TOKEN / API_KEY / 不改。
func PickAuth(t *Terminal, current string) string {
	items := []string{i18n.T("pick.auth.token"), i18n.T("pick.auth.apikey"), i18n.T("pick.noChange")}
	s := SelectMenu(t, SelectOptions{Title: i18n.T("pick.auth.title", current), Items: items, Hint: i18n.T("pick.hint"), NoNumber: true})
	if s == 0 {
		return presets.AuthToken
	}
	if s == 1 {
		return presets.AuthAPIKey
	}
	return current
}

var effortOpts = []string{"low", "medium", "high", "xhigh", "max", "auto"}

// PickEffort 选 effort 思考档：low…auto / 留空 / 不改。
func PickEffort(t *Terminal, current string) string {
	items := append(append([]string{}, effortOpts...), i18n.T("pick.effort.empty"), i18n.T("pick.noChange"))
	s := SelectMenu(t, SelectOptions{Title: i18n.T("pick.effort.title", orEmpty(current)), Items: items, Hint: i18n.T("pick.effort.hint"), NoNumber: true})
	if s < 0 || s == len(items)-1 {
		return current
	}
	if s == len(effortOpts) {
		return ""
	}
	return effortOpts[s]
}
